### DISCUSSION TITLE
I want to use Opal as node.js syntax sugaring. Doe it support my use case?

### DISCUSSION THREAD

**Discussion by midnight-wonderer** [opal/opal#2692] [Q&A]
Opal's document is not comprehensive, so I can't decide whether it supports my use case.
I want something far simpler than what the tutorial shows. I want to write a pure Ruby class that can interact with node.js code.

### Here is the list of the features I need.
- Ruby class compiles into Javascript class.
- The compiled class must be able to be initialized from Javascript.
- All methods of the compiled class are declared as async.
- I can pass Javascript functions into the Ruby class.
- The function passed into the class should be able to get called from the Ruby side, assuming that all functions are async functions.
- Only the minimum Ruby runtime is needed. I do not plan to `require` anything in the Ruby files. However, integer, boolean, object, hash, array operations, and maybe some other basic stuff should work. I am not looking for a full-blown Ruby runtime like `ostruct`, for instance.
- Ruby flow controls, including error handling.
- Ruby `Error` interop with Javascript. The Javascript side should be able to catch the error raised from the Ruby side and vice versa.
- This implied that there should be a good interface for importing Javascript error classes to be used in the Ruby `rescue` syntax and also exporting Ruby error classes to be used with `instanceof` on the Javascript side.
- Sting, Number, Boolean, Hash, and Array interopts. (Basically, everything that can be represented in JSON.)
- The input is multiple Ruby class files; the result should be multiple Javascript files with shared runtime. (i.e., not duplicating the runtime on each file.)

I know, by trying out what is in the README, that Opal accomplished much more than this, but I am unsure if my use case is supported or even possible.

### Questions
- Is Opal suitable for my use case?
- If Opal does not support the use case, do you know alternatives? (If they existed.)
- I would appreciate any comment shedding light on how impossible my requirements are. (As someone who does not develop compilers or interpreters, I don't know the depth.)

**Comment by hmdne** [comment]
> Opal's document is not comprehensive, so I can't decide whether it supports my use case. I want something far simpler than what the tutorial shows. I want to write a pure Ruby class that can interact with node.js code.

Let me try to answer your questions :)

> ### Here is the list of the features I need.
> 
>     * Ruby class compiles into Javascript class.

Yes, Ruby classes are represented as JavaScript classes via the prototype system.

> 
>     * The compiled class must be able to be initialized from Javascript.

No problem, but you need to use a slightly different syntax, eg.:

```js
# For MyNameSpace::MyKlass:
instance = Opal.MyNameSpace.MyKlass.$new() # perhaps with arguments, block passing will be harder
```

> 
>     * All methods of the compiled class are declared as async.

That's not the case with Opal. But - we support `async` and `await`:

```ruby
# await: true

def this_is_an_async_function
  nil.__await__
end
```

If `__await__` keyword is enabled (via a magic comment) and it's used anywhere in a function (or a block), this function (block) becomes async. Do note that this exceeds Ruby semantics a little, so a code like:

```ruby
(1..10).map { this_is_an_async_function }
```

This code as well (this will compile a block to become async):

```ruby
(1..10).map { this_is_an_async_function.__await__ }
```

Will return a map of Promises, not nils. There are a couple of helper functions in `await` file in Opal stdlib for such cases. Ensure you read documentation on that topic though: https://github.com/opal/opal/blob/master/docs/async.md

> 
>     * I can pass Javascript functions into the Ruby class.

Not an issue, except for perhaps the `self` semantics which needs to be emulated. But since you want to call from JS, you expect JS semantics.

Passing block arguments (unlike simple `Proc` arguments) can be a little tricky though:

```ruby
instance.accepts_block { |s| s*2 }
```

is equivalent to:

```js
instance.$accepts_block.$$p = function(s) { return s*2; };
instance.$accepts_block();
```

In any case, you can check the generated code with the following command:
```bash
opal -Oce "instance.accepts_block { |s| s*2 }"
```
(this in particular will use a helper function Opal.send, which is basically equivalent to the above)

> 
>     * The function passed into the class should be able to get called from the Ruby side, assuming that all functions are async functions.
Not all functions are async functions, but you can certainly do something like

```ruby
# await: true

def some_function_accepting_js_async_function(js_func)
  js_func.call.__await__
end
```

(Note, that on Ruby side, JS functions will be of a class `Proc`)

> 
>     * Only the minimum Ruby runtime is needed. I do not plan to `require` anything in the Ruby files. However, integer, boolean, object, hash, array operations, and maybe some other basic stuff should work. I am not looking for a full-blown Ruby runtime like `ostruct`, for instance.

You can manually select parts of Opal. The default is to use `corelib`, and `ostruct` isn't part of that (it's part of `stdlib`). You can even choose to include a smaller part of Opal, by requiring (on client side) `opal/mini` or `opal/core` instead of `opal` (note that likely your programs won't run, unless you are extra careful).

> 
>     * Ruby flow controls, including error handling.

Opal supports that. Most cases of Ruby control flow are supported. If some special case doesn't work, that's a bug.

> 
>     * Ruby `Error` interop with Javascript. The Javascript side should be able to catch the error raised from the Ruby side and vice versa.

Yes, but - on Ruby side, you will need to explicitly catch `Opal::Raw::Error` if a JS error class isn't bridged (just like you need to catch `Exception`).

> 
>     * This implied that there should be a good interface for importing Javascript error classes to be used in the Ruby `rescue` syntax and also exporting Ruby error classes to be used with `instanceof` on the Javascript side.

It is possible to bridge other types from JS to Ruby but it's not recommended in general. But it's documented: https://github.com/opal/opal/blob/master/docs/bridging.md

On the reverse side, you can always use `Opal.A.B.C` to refer to `A::B::C` class.

> 
>     * Sting, Number, Boolean, Hash, and Array interopts. (Basically, everything that can be represented in JSON.)

There are bridges for:

* JS String -> Ruby `String`
* JS `Number` -> Ruby `Number` (a special class that encompases both `Float`s and `Integer`s)
* JS `Boolean` -> Ruby `TrueClass`/`FalseClass
* JS `Map` -> Ruby `Hash`
* JS `Array` -> Ruby `Array`
* JS `Function` -> Ruby `Proc`
* JS `Promise` -> Ruby `PromiseV2` (you will need to `require "promise/v2"` from our stdlib to have that)

Note that the following are not bridged and will generate incorrect code if you are not careful:

`null`, `undefined`, plain JS object (ie. `{a: 4}`).

If on Ruby side you encounter an incorrect object, you eg. won't be able to call methods on that object. You can always interpolate JavaScript using the backtick operator (you will also need a magic comment), or use the special #JS method.

```ruby
# backtick_javascript: true

# this function is given an argument of {a: 4}
def accepts_plain_js_object(obj)
  `obj.a` # => 4
  obj.JS.a # => 4
end
```

You can also take a look at our corelib (it's written in hybrid Ruby and JavaScript), since your code will likely look similar: https://github.com/opal/opal/blob/master/opal/corelib/method.rb

There's also an `stdlib` `Opal::Raw` class which will allow you to use Ruby for calls like instanceof:

https://github.com/opal/opal/blob/master/stdlib/opal/raw.rb

I'd also recommend you read the following documentation:

https://github.com/opal/opal/blob/master/docs/compiled_ruby.md

> 
>     * The input is multiple Ruby class files; the result should be multiple Javascript files with shared runtime. (i.e., not duplicating the runtime on each file.)

It's called directory mode, but it's not released (on `master`). With released Opal version, you can compile all Ruby files into one JS file, then depending on what you want, call a JS function `Opal.require("file1")` which will execute a given file. You can also separately compile runtime and each file, then load runtime+file1, or runtime+file2.

For the last case, an example on how to achieve that is in Opal-Browser readme:

https://github.com/opal/opal-browser

> 
> 
> I know, by trying out what is in the README, that Opal accomplished much more than this, but I am unsure if my use case is supported or even possible.
> ### Questions
> 
>     * Is Opal suitable for my use case?

I think yes, but I feel you may need to write some complex code when dealing with interoperability, so you will likely want to abstract that. If it's possible, then especially the `async` code which is ugly in Opal.

The best idea would be to write your project (library, I assume) in Ruby, then add JavaScript boilerplate for that.

> 
>     * If Opal does not support the use case, do you know alternatives? (If they existed.)

I am aware of RubyJS project, which basically compiles the Ruby syntax into JS syntax. Some methods are translated, but don't expect having a fully working Ruby runtime.

> 
>     * I would appreciate any comment shedding light on how impossible my requirements are. (As someone who does not develop compilers or interpreters, I don't know the depth.)

I'd say, Opal compiler is not that complicated. We don't do Ruby parsing, we just deal with the syntax tree produced by the `parser` Gem, then convert that tree step by step into compatible JavaScript. With `corelib` it should produce a working program. We use similarities between JS and Ruby, but where the languages are dissimilar (eg. in `async`/`await` case) there may happen to be issues. If you are interested, I highly recommend you read our source code :)

* `lib` contains the compiler (and other functions needed for host-side)
* `opal/corelib` contains the corelib (most of that is loaded by default, but some are not, so you must `require "corelib/pattern_matching"` if you wanted support for pattern matching)
* `stdlib` contains the stdlib (helper classes like `Opal::Raw` or other libraries like `OpenStruct`)
